File: 2026-01-22-user-sorting/java/Solution.java
import java.util.*;
import java.util.stream.*;

class User {
    String firstName;
    String lastName;
    int age;
    String email;

    User(String firstName, String lastName, int age, String email) {
        this.firstName = firstName;
        this.lastName = lastName;
        this.age = age;
        this.email = email;
    }
}

class UserResult {
    String fullName;
    String email;

    UserResult(String fullName, String email) {
        this.fullName = fullName;
        this.email = email;
    }
}

public class Solution {
    public static List<UserResult> processUsers(List<User> users) {
        return users.stream()
                .filter(u -> u.age >= 18)
                .sorted(Comparator.comparing(u -> u.lastName))
                .map(u -> new UserResult(u.firstName + " " + u.lastName, u.email))
                .collect(Collectors.toList());
    }

    public static void main(String[] args) {
        List<User> users = List.of(
            new User("Alice", "Zuberg", 25, "alice@example.com"),
            new User("Bob", "Alpha", 17, "bob@example.com"),
            new User("Charlie", "Beta", 30, "charlie@example.com")
        );

        processUsers(users).forEach(
            u -> System.out.println(u.fullName + " - " + u.email)
        );
    }
}

File: 2026-01-22-user-sorting/python/solution.py
def process_users(users):
    filtered = [u for u in users if u["age"] >= 18]
    sorted_users = sorted(filtered, key=lambda u: u["lastName"])
    return [
        {
            "fullName": f"{u['firstName']} {u['lastName']}",
            "email": u["email"]
        }
        for u in sorted_users
    ]


users = [
    {"firstName": "Alice", "lastName": "Zuberg", "age": 25, "email": "alice@example.com"},
    {"firstName": "Bob", "lastName": "Alpha", "age": 17, "email": "bob@example.com"},
    {"firstName": "Charlie", "lastName": "Beta", "age": 30, "email": "charlie@example.com"}
]

for user in process_users(users):
    print(user)
